javalib/src/main/scala/java/net/HttpCookie.scala

```diff
@@ -0,0 +1,413 @@
+package java.net
+
+import java.util.ArrayList
+
+object HttpCookie {
+
+  def domainMatches(domain: String, host: String): Boolean = {
+    val effectiveHost = if (host.indexOf('.') == -1) {
+      host + ".local"
+    } else
+      host
+
+    val domainL = domain.toLowerCase
+    val hostL = host.toLowerCase
+
+    val exactMatch = domainL == hostL
+    // a host is a subdomain if it ends with the domain, and
+    // the character preceeding the domain suffix is a dot
+    val hostIsSubdomain = hostL.endsWith("." + domainL)
+
+    domainL == hostL || hostIsSubdomain
+
+  }
+
+  private final val Discard = "Discard"
+  private final val Comment = "Comment"
+  private final val CommentURL = "CommentURL"
+  private final val Domain = "Domain"
+  private final val MaxAge = "Max-Age"
+  private final val Path = "Path"
+  private final val Port = "Port"
+  private final val Secure = "Secure"
+  private final val Version = "Version"
+  private final val HttpOnly = "HttpOnly"
+
+  def parse(header: String): java.util.List[HttpCookie] = {
+
+    val cookies = new ArrayList[HttpCookie]()
+
+    if (header == null)
+      throw new NullPointerException()
+    else if (header == "")
+      cookies
+    else {
+
+      val nameless =
+        if (header.regionMatches(false, 0, "set-cookie2:", 0, 12)) {
+          header.substring(12)
+        } else if (header.regionMatches(false, 0, "set-cookie:", 0, 11))
+          header.substring(11)
+        else
+          header
+
+      val unprocessed = splitUnquotedAt(nameless, ',')
+
+      var i = 0
+      while (i < unprocessed.size()) {
+        val kv = unprocessed.get(i)
+        val eqIndex = kv.indexOf("=")
+        val name = kv.substring(0, eqIndex)
+        val value = kv.substring(eqIndex + 1)
+
+        val cookie = parseValue(name.trim(), unquote(value.trim()))
+        if (cookie != null) {
+          cookies.add(cookie)
+        }
+
+        i += 1
+
+      }
+
+    }
+
+    cookies
+
+  }
+
+  // Splits the string at the given character,
+  // however it also respects quoted strings, i.e.
+  // occurences of the separator inside quoted strings
+  // won't trigger a split
+  // backslash also escapes double quotes inside the quoted sections
+  private def splitUnquotedAt(
+      in: String,
+      separator: Char
+  ): java.util.List[String] = {
+
+    val res = new ArrayList[String]()
+    if (in == "") {
+      res
+    } else {
+
+      var i = 0
+      var begin = 0
+      var inQuotes = false
+      var escaping = false
+
+      while (i < in.length()) {
+
+        val char = in.charAt(i)
+
+        if (char == separator && !inQuotes) {
+          res.add(in.substring(begin, i))
+          begin = i + 1
+        } else if (char == '\\' && inQuotes && !escaping) {
+          escaping = true
+        } else if (char == '"' && inQuotes && !escaping) {
+          inQuotes = false
+        } else if (char == '"' && !inQuotes) {
+          inQuotes = true
+        }
```

Comment:
How about refactoring `inQuotes && !escaping` into a common condition so we don't need to check it twice? 
```suggestion
        } else if (inQuotes){
         if(!escaping) char match {
           case '\\' =>  escaping = true
           case '"'  =>  inQuotes = false
           case _ => ()
        }
        else if (char == '"') inQuotes = true
      
```